src/runtime/refiner.ts

```diff
@@ -1237,143 +1241,220 @@ export class Fraction {
-    return new Polynomial(pn.coeffs, pn.indeterminate);
+  static fromKey(key: string): Term {
+    return new Term(JSON.parse(key));
   }
 
-  get coeffs(): number[] {
-    while (this._coeffs.length >= 1 && this._coeffs[this._coeffs.length - 1] === 0) {
-      this._coeffs.pop();
+  static indeterminateToExpression(fn: string, pow: number): RefinementExpression {
+    if (pow <= 0) {
+      throw new Error('Must have positive power.');
     }
-    if (this._coeffs.length === 0) {
-      this._coeffs.push(0);
+    if (pow === 1) {
+      return new FieldNamePrimitive(fn, Primitive.NUMBER);
     }
-    return this._coeffs;
+    return new BinaryExpression(
+      Term.indeterminateToExpression(fn, 1),
+      Term.indeterminateToExpression(fn, pow - 1),
+      new RefinementOperator(Op.MUL));
   }
 
-  set coeffs(cfs: number[]) {
-    this._coeffs = [...cfs];
+  // assumes that term is not a constant i.e. not {}
+  toExpression(): RefinementExpression {
+    if (Object.keys(this.indeterminates).length === 0) {
+      throw new Error('Cannot convert an empty term to expression');
+    }
+    let expr = null;
+    for (const [indeterminate, power] of Object.entries(this.indeterminates)) {
+      const indtrExpr = Term.indeterminateToExpression(indeterminate, power);
+      expr = expr ? new BinaryExpression(expr, indtrExpr, new RefinementOperator(Op.MUL)) : indtrExpr;
+    }
+    return expr;
   }
+}
 
-  degree(): number {
-    return this.coeffs.length - 1;
+export class Multinomial {
+  private _terms: Dictionary<number>;
+
+  constructor(terms: Dictionary<number> = {}) {
+    this.terms = terms;
+  }
+
+  static copyOf(mn: Multinomial): Multinomial {
+    return new Multinomial(mn.terms);
   }
 
-  static assertCompatibility(a: Polynomial, b: Polynomial) {
-    if (a.indeterminate && b.indeterminate && a.indeterminate !== b.indeterminate) {
-      throw new Error('Incompatible polynomials');
+  get terms(): Dictionary<number> {
+    for (const [term, coeff] of Object.entries(this._terms)) {
+      if (coeff === 0) {
+        delete this._terms[term];
+      }
     }
+    const ordered = {};
+    const unordered = this._terms;
+    Object.keys(unordered).sort().forEach((key) => {
+      ordered[key] = unordered[key];
+    });
+    this._terms = ordered;
+    return this._terms;
+  }
+
+  set terms(tms: Dictionary<number>) {
+    this._terms = {...tms};
   }
 
-  static add(a: Polynomial, b: Polynomial): Polynomial {
-    Polynomial.assertCompatibility(a, b);
-    const sum = a.degree() > b.degree() ? Polynomial.copyOf(a) : Polynomial.copyOf(b);
-    const other = a.degree() > b.degree() ? b : a;
-    for (const [i, coeff] of other.coeffs.entries()) {
-      sum.coeffs[i] += coeff;
+  static add(a: Multinomial, b: Multinomial): Multinomial {
+    const sum = new Multinomial();
+    for (const [term, coeff] of Object.entries(a.terms)) {
+      const val = coeff + (sum.terms.hasOwnProperty(term) ? sum.terms[term] : 0);
```

Comment:
I think the following is a bit simpler (and still safe as the only falsey value that is allowed in the terms is a zero)
```
      const val = coeff + sum.terms[term] || 0;
```

Reply:
In fact I think the first loop can be simplified to a copy.
```
const sum = a.copyOf()
```
and the second
```
for (const [term, coeff] of Object.entries(b.terms)) {
    sum.terms[term] = sum.terms[terms] || 0 + coeff;
}
```

Reply:
Oo a lot neater.

Discovered 2 more things:

* Need to put brackets in order to make it work. Operator precedence stuff.
```
const val = (sum.terms[terms] || 0) + coeff;
```
* Apparently you cannot avoid updating the sum as a 2 step process i.e. 
```
const val = (sum.terms[term] || 0) + coeff;
sum.terms[term] = val;
```
The following does not work:
```
sum.terms[term] = sum.terms[terms] || 0 + coeff;
```